### PR TITLE
dnsmasq: listen to DHCP requests on secondary networks too

### PR DESCRIPTION
**Pull request by jean-edouard**
Also, start mac addresses on secondary networks at 1 instead of 0, to match the primary network

### CODE REVIEW

**Review by ormergi — COMMENTED** [review]
Hi Jed, Could you please elaborate why we need those changes on the PR description and on commit message?

File: cluster-provision/centos8/scripts/dnsmasq.sh
Comment (jean-edouard) [reply cluster-provision/centos8/scripts/dnsmasq.sh:36]: Well more interfaces might get added in the for loop below, hence the use of plural.

File: cluster-provision/centos8/scripts/dnsmasq.sh
Comment (jean-edouard) [reply cluster-provision/centos8/scripts/dnsmasq.sh:42]: No, it does not mean secondary networks are connected to the primary one.
The link between node NICs doesn't happen here.
This also makes sure every network uses a different IP range.

File: cluster-provision/gocli/cmd/run.go
Comment (jean-edouard) [reply cluster-provision/gocli/cmd/run.go:435]: This is not strictly necessary (anymore), since I don't set `--dhcp-host` for nodes on secondary networks, since they don't need an IP on those.
But I left it there for consistency:
On the primary network, this number is the node number. Without my change, on secondary networks, this number is the node number - 1.

**Comment by jean-edouard** [comment]
Thanks for the review @ormergi.
I know the PR description and commit message might seem vague, this is because I didn't want to tie this change, and https://github.com/kubevirt/kubevirtci/pull/651, to a particular feature.
Secondary networks are there to help us test more use cases, and being able (not required) to use DHCP on those further expands the possibilities.

I did write this to enable writing tests for a specific feature: dedicated migration networks (https://github.com/kubevirt/kubevirt/pull/6036).
The feature documentation will state that the cluster admin needs to setup a physical secondary network with a DHCP server running on it.
My 2 PRs enable emulating that scenario.

**Comment by rmohr** [comment]
/approve

Looks good to me. Letting @ormergi do the lgtm once he is satisfied.

**Comment by ormergi** [comment]
> Thanks for the review @ormergi.
> I know the PR description and commit message might seem vague, this is because I didn't want to tie this change, and #651, to a particular feature.
> Secondary networks are there to help us test more use cases, and being able (not required) to use DHCP on those further expands the possibilities.
> 
> I did write this to enable writing tests for a specific feature: dedicated migration networks ([kubevirt/kubevirt#6036](https://github.com/kubevirt/kubevirt/pull/6036)).
> The feature documentation will state that the cluster admin needs to setup a physical secondary network with a DHCP server running on it.
> My 2 PRs enable emulating that scenario.

I understand, I think it would be much easier to grasp if you would right that on the PR description and commit message

**Review by ormergi — APPROVED** [review]
Thanks!

**Review by EdDev — COMMENTED** [review]
@jean-edouard , enabling dynamic IP on secondary networks means we may get multiple default gateways and dns entries on the nodes, which may break its connectivity randomly.

I pretty much forgot how this looks on the nodes. Do they also get an IP on the other nics?
Could you please share the nodes and a VMI with secondary networks look like now? I am mainly interested in the addresses and routes (ip address, ip route).

/hold

**Comment by jean-edouard** [comment]
> @jean-edouard , enabling dynamic IP on secondary networks means we may get multiple default gateways and dns entries on the nodes, which may break its connectivity randomly.
> 
> I pretty much forgot how this looks on the nodes. Do they also get an IP on the other nics?

They do not, and indeed we don't want them to.

> Could you please share the nodes and a VMI with secondary networks look like now? I am mainly interested in the addresses and routes (ip address, ip route).

I haven't used the secondary network in VMIs. I only link virt-handlers to that secondary network for migration purposes.
Since nodes don't dhclient on that secondary network, routes are unchanged.

> 
> /hold

**Comment by EdDev** [comment]
> I haven't used the secondary network in VMIs. I only link virt-handlers to that secondary network for migration purposes.

ok, but I am worried about the existing network tests that may assume no dhcp/ra server exists on these.
I want to make sure that if we connect a secondary interface to the vmi, it will not by default enable that interface with dynamic IP. I can check it on Sunday, if you are ok waiting until then.

File: cluster-provision/centos8/scripts/dnsmasq.sh
Comment (phoracek) [inline cluster-provision/centos8/scripts/dnsmasq.sh:51]: With this change, we may need to force removal of secondary network profiles on nodes (https://github.com/nmstate/kubernetes-nmstate/blob/main/cluster/up.sh#L14). Otherwise, nodes may start up inconsistently with different default network each time they boot.

Comment (EdDev) [reply cluster-provision/centos8/scripts/dnsmasq.sh:51]: > > I haven't used the secondary network in VMIs. I only link virt-handlers to that secondary network for migration purposes.
> 
> ok, but I am worried about the existing network tests that may assume no dhcp/ra server exists on these.
> I want to make sure that if we connect a secondary interface to the vmi, it will not by default enable that interface with dynamic IP. I can check it on Sunday, if you are ok waiting until then.

@ormergi has checked this one (Or, please correct me here).
If this the server is set for secondary networks, with a NAD that just defines a bridge CNI with no IPAM, the VMI guests we use will automatically get an IP & routes because this is NetworkManager default.

If this is indeed correct, we have a problem (even though the current tests somehow pass).
Just because without doing any configuration on the guests (usually through cloud-init-network), we will end up with addresses deployed there in secondary interfaces.

An alternative could be to dedicate a specific secondary network for this migration network and not enable the dnsmasq on all. WDYT?

**Comment by jean-edouard** [comment]
Closing.
We'll use `whereabouts` for testing dedicated migration networks, so this is not needed anymore.

**Pull request closed without merging**
